lib/auth/join/boundkeypair/boundkeypair.go

```diff
@@ -158,46 +362,82 @@ func LoadClientState(ctx context.Context, fs FS) (*ClientState, error) {
 		return nil, trace.Wrap(err, "parsing private key")
 	}
 
+	var keyHistory []KeyHistoryEntry
+	keyHistoryBytes, err := fs.Read(ctx, KeyHistoryPath)
+	if trace.IsNotFound(err) {
+		// No history, this is allowed.
+	} else if err != nil {
+		slog.WarnContext(ctx, "unable to read key history, may be unable to recover in the event of a cluster rollback", "error", err)
+	} else if len(keyHistoryBytes) > 0 {
+		keyHistory, err = parseKeyHistory(keyHistoryBytes)
+		if err != nil {
+			slog.WarnContext(ctx, "unable to parse key history, may be unable to recover in the event of a cluster rollback", "error", err)
+		}
+	}
+
+	// If the key history is empty, initialize it with just the current key.
+	if len(keyHistory) == 0 {
+		keyHistory = []KeyHistoryEntry{{
+			Time:       time.Now(),
+			PrivateKey: string(privateKeyBytes),
+		}}
+	}
+
 	return &ClientState{
-		PrivateKey: pk,
+		fs: fs,
 
+		PrivateKey:      pk,
 		PrivateKeyBytes: privateKeyBytes,
 		JoinStateBytes:  joinStateBytes,
+		KeyHistory:      keyHistory,
 	}, nil
 }
 
 // StoreClientState writes bound keypair client state to the given filesystem
 // wrapper. Public keys and join state will only be written if
-func StoreClientState(ctx context.Context, fs FS, state *ClientState) error {
-	if err := fs.Write(ctx, PrivateKeyPath, state.PrivateKeyBytes); err != nil {
+func (c *ClientState) Store(ctx context.Context) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if err := c.fs.Write(ctx, PrivateKeyPath, c.PrivateKeyBytes); err != nil {
 		return trace.Wrap(err, "writing private key")
 	}
 
-	// TODO: maybe consider just not writing the public key at all. End users
-	// aren't really meant to look in the internal storage, and we can just
-	// derive the public key whenever we want.
-
 	// Only write the public key if it was explicitly provided. This helps save
 	// an unnecessary file write.
-	if len(state.PublicKeyBytes) > 0 {
-		if err := fs.Write(ctx, PublicKeyPath, state.PublicKeyBytes); err != nil {
+	if len(c.PublicKeyBytes) > 0 {
+		if err := c.fs.Write(ctx, PublicKeyPath, c.PublicKeyBytes); err != nil {
 			return trace.Wrap(err, "writing public key")
 		}
 	}
 
-	if len(state.JoinStateBytes) > 0 {
-		if err := fs.Write(ctx, JoinStatePath, state.JoinStateBytes); err != nil {
+	if len(c.JoinStateBytes) > 0 {
+		if err := c.fs.Write(ctx, JoinStatePath, c.JoinStateBytes); err != nil {
 			return trace.Wrap(err, "writing previous join state")
 		}
 	}
 
+	if len(c.KeyHistory) > 0 {
+		// Keep only the first N elements to avoid excessive lookup times.
+		bytes, err := json.Marshal(c.KeyHistory[:min(len(c.KeyHistory), KeyHistoryLength)])
```

Comment:
As we're already doing this trimming in `GenerateKeypair`, is this necessary?

Reply:
Fair point, I've removed the extra trimming in [375bc16](https://github.com/gravitational/teleport/pull/55240/commits/375bc160d6b659c8709ba62e0fe5ad0f6372693e)